File: backend-fastapi/app/services/attendance_overstay_scheduler.py
import logging
from datetime import datetime

from sqlalchemy import select

from app.core.database import SessionLocal
from app.models.attendance import Attendance
from app.models.fitness_user import FitnessUser
from app.models.membership import MembershipPlan
from app.models.role import Role
from app.models.users import Users
from app.services import notification_service

log = logging.getLogger("fitnexus.attendance_overstay")


def check_overstays() -> None:
    with SessionLocal() as db:
        open_sessions = list(db.scalars(select(Attendance).where(Attendance.check_out_time.is_(None))))
        for record in open_sessions:
            try:
                _check_one(db, record)
            except Exception:
                log.warning("Overstay check failed for attendance %s", record.id, exc_info=True)


def _check_one(db, record: Attendance) -> None:
    if record.overstay_notified or record.check_in_time is None:
        return
    member = db.get(Users, record.user_id)
    if member is None or Role(member.role) != Role.USER:
        return
    fu = db.get(FitnessUser, member.id)
    plan = db.get(MembershipPlan, fu.membership_plan_id) if fu and fu.membership_plan_id else None
    limit = plan.max_session_minutes if plan else None
    if not limit or limit <= 0:
        return

    minutes = (datetime.now() - record.check_in_time).total_seconds() / 60
    if minutes < limit:
        return

    plan_name = plan.name if plan else "current"
    recipients = {uid for uid in (member.admin_id,) if uid}
    if not recipients:
        recipients = set(
            db.scalars(select(Users.id).where(Users.role.in_([Role.ADMIN.value, Role.SUPER_ADMIN.value])))
        )
    for rid in recipients:
        notification_service.create_for_user(
            db, rid, "INFO", "Member overstay",
            f"{member.name} has been in the gym {int(minutes)} min — over the {limit}-min limit of the "
            f"{plan_name} plan — and hasn't checked out.",
            "/attendance",
        )
    notification_service.create_for_user(
        db, member.id, "INFO", "Session time exceeded",
        f"You've passed your {plan_name} plan's {limit}-minute visit limit. Upgrade your plan to extend "
        "your time at the gym.",
        "/membership",
    )

    record.overstay_notified = True
    record.updated_at = datetime.now()
    db.commit()

File: backend/src/main/java/com/fitnexus/backend/config/AttendanceOverstayScheduler.java
package com.fitnexus.backend.config;

import com.fitnexus.backend.entity.Attendance;
import com.fitnexus.backend.entity.FitnessUser;
import com.fitnexus.backend.entity.MembershipPlan;
import com.fitnexus.backend.entity.Role;
import com.fitnexus.backend.entity.Users;
import com.fitnexus.backend.repository.AttendanceRepository;
import com.fitnexus.backend.repository.FitnessUserRepository;
import com.fitnexus.backend.repository.UserRepository;
import com.fitnexus.backend.service.NotificationService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.stream.Stream;

/**
 * Periodically scans members who are still inside the gym and flags anyone who has
 * stayed past their plan's per-visit limit (e.g. Basic = 1 hour). On a first breach
 * it notifies the member's admin ("not exited yet") and the member themselves
 * ("upgrade for longer access"). Premium / unlimited plans are never flagged.
 */
@Component
@RequiredArgsConstructor
@Slf4j
public class AttendanceOverstayScheduler {

    private final AttendanceRepository attendanceRepository;
    private final FitnessUserRepository fitnessUserRepository;
    private final UserRepository userRepository;
    private final NotificationService notificationService;

    /** Runs every 5 minutes by default (configurable), after a 1-minute startup delay. */
    @Scheduled(fixedDelayString = "${attendance.overstay-check-ms:300000}", initialDelay = 60000)
    @Transactional
    public void flagOverstays() {
        LocalDateTime now = LocalDateTime.now();
        for (Attendance a : attendanceRepository.findByCheckOutTimeIsNull()) {
            try {
                if (Boolean.TRUE.equals(a.getOverstayNotified())) {
                    continue;
                }
                Users member = a.getUser();
                if (member == null || member.getRole() != Role.USER || a.getCheckInTime() == null) {
                    continue;
                }
                FitnessUser fu = fitnessUserRepository.findById(member.getId()).orElse(null);
                MembershipPlan plan = fu != null ? fu.getMembershipPlanRef() : null;
                Integer limit = plan != null ? plan.getMaxSessionMinutes() : null;
                if (limit == null || limit <= 0) {
                    continue; // unlimited session — no overstay
                }
                long minutes = Duration.between(a.getCheckInTime(), now).toMinutes();
                if (minutes < limit) {
                    continue;
                }
                String planName = plan.getName() != null ? plan.getName() : "your";
                notifyAdmins(member, planName, limit, minutes);
                notificationService.createForUser(member, "ATTENDANCE", "Session time exceeded",
                        "You've passed your " + planName + " plan's " + limit
                                + "-minute visit limit. Upgrade your plan to extend your time at the gym.",
                        "/membership");
                a.setOverstayNotified(true);
                a.setUpdatedAt(now);
                attendanceRepository.save(a);
            } catch (Exception e) {
                log.warn("Overstay check failed for attendance {}", a.getId(), e);
            }
        }
    }

    private void notifyAdmins(Users member, String planName, int limit, long minutes) {
        String title = "Member overstay";
        String message = member.getName() + " has been in the gym " + minutes + " min — over the "
                + limit + "-min limit of the " + planName + " plan — and hasn't checked out.";
        Users admin = member.getAdmin();
        if (admin != null) {
            notificationService.createForUser(admin, "ATTENDANCE", title, message, "/attendance");
            return;
        }
        // No linked admin — fall back to all admins / super admins.
        Stream.concat(userRepository.findByRole(Role.ADMIN).stream(),
                        userRepository.findByRole(Role.SUPER_ADMIN).stream())
                .forEach(a -> notificationService.createForUser(a, "ATTENDANCE", title, message, "/attendance"));
    }
}
